### PR TITLE
Jar Assembly Support

### PR DESCRIPTION
**Pull request by Milyardo**
Hack in support for Jars on the classpath by extracting them to a temp directory.

Perhaps extracting to hidden directory in user.home might be better. Perhaps a sbt task to modify the classpath, replacing jars with unzipped contents contents would an option as well.

### CODE REVIEW

**Comment by densh** [comment]
Hi @Milyardo , thanks a lot for looking at this issue!

Unfortunately the approach you've taken has an major issue that needs to be addressed before it gets in: any I/O during linking is expensive, so we should really not touch disk unless we absolutely have to. For example current implementation of directory assembly loader right now never opens files unless someone has to be loaded from it, just traverses and indexes directory structure.

We need to do something similar here with jar support: 1. we can not expand jars to local folders as that can be quite slow. 2. instead we need to lazily load NIR files and create in-memory ByteBuffer-s for them lazily whenever a certain entry is accessed, but not sooner than that.

**Comment by densh** [comment]
Also please rebase on top of latest master, especially after @andreaTP's changes in https://github.com/scala-native/scala-native/pull/94 get merged.

**Comment by sjrd** [comment]
This is especially important on Windows. We had huge performance problems when we were extracting .jars on disk on Windows. And *not* because of extraction (that was cached), but because *reading* all these tiny files was really really slow.

It was all fixed by unzipping the jars in memory.

**Comment by Milyardo** [comment]
Sure thing @densh, I was already poking at an in memory solution, I'll update this PR something this evening.

**Comment by sjrd** [comment]
If you want inspiration, we have some pretty well battle-tested abstractions for files on disk/in memory, within .jar files or not, etc. at https://github.com/scala-js/scala-js/tree/master/tools/shared/src/main/scala/org/scalajs/core/tools/io and https://github.com/scala-js/scala-js/tree/master/tools/jvm/src/main/scala/org/scalajs/core/tools/io

**Comment by Milyardo** [comment]
Here's an simple Hello world project linking to clib: https://github.com/Milyardo/scalaNativeHelloWorld

File: tools/src/main/scala/scala/scalanative/io/Classpath.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import java.net.URI
+import java.nio.ByteBuffer
+import java.nio.file._
+
+import scala.collection.JavaConverters._
+
+/**
+  * A values which represents a member of a classPath.
+  */
+sealed abstract class Classpath[T <: VFile] {
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/io/Classpath.scala]: Given that `Classpath` is only used by linker, it should stay in `linker` package.

File: tools/src/main/scala/scala/scalanative/io/Classpath.scala [hunk 2]
```diff
@@ -0,0 +1,55 @@
+package scala.scalanative.io
+
+import java.net.URI
+import java.nio.ByteBuffer
+import java.nio.file._
+
+import scala.collection.JavaConverters._
+
+/**
+  * A values which represents a member of a classPath.
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/io/Classpath.scala]: s/classPath/classpath/g

File: tools/src/main/scala/scala/scalanative/io/Classpath.scala [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  */
+sealed abstract class Classpath[T <: VFile] {
+  def contents: Stream[T]
+  def close(): Unit
+}
+
+/**
+  * A trait for abstracting memory cached file-like object contained in some member of a classpath.
+  */
+trait VFile {
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/io/Classpath.scala]: s/VFile/VirtualFile/g

File: tools/src/main/scala/scala/scalanative/io/Classpath.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  override def contents: Stream[LazyVFile[JarClasspath]] = {
+    roots
+      .flatMap(path => Files.walk(path, Integer.MAX_VALUE, FileVisitOption.FOLLOW_LINKS).iterator().asScala)
+      .map(path => LazyVFile[JarClasspath](path, this))
+  }
+
+  override def close(): Unit = fs.close()
+}
+
+final case class LazyVFile[T <: Classpath[LazyVFile[T]]](relativePath: Path, fs: T) extends VFile {
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/io/Classpath.scala]: s/LazyVFile/LazyVirtualFile/g

File: tools/src/main/scala/scala/scalanative/io/Classpath.scala [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    .iterator()
+    .asScala.toStream
+    .map(filePath => LazyVFile[LocalClasspath](filePath, this))
+}
+
+final case class JarClasspath(path: Path) extends Classpath[LazyVFile[JarClasspath]] {
+  val fsUri = URI.create(s"jar:${path.toUri}")
+  val options = Map("create" -> "false")
+  val fs: FileSystem = FileSystems.newFileSystem(fsUri, options.asJava)
+  val roots: Stream[Path] = fs.getRootDirectories.asScala.toStream
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/io/Classpath.scala]: All of the vals above should be private.

File: tools/src/main/scala/scala/scalanative/io/Classpath.scala [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+/**
+  * A trait for abstracting memory cached file-like object contained in some member of a classpath.
+  */
+trait VFile {
+  def relativePath: Path
+  def bytes: ByteBuffer
+  def uri: URI
+}
+
+final case class LocalClasspath(path: Path) extends Classpath[LazyVFile[LocalClasspath]] {
```
Comment (sjrd) [inline tools/src/main/scala/scala/scalanative/io/Classpath.scala]: If you don't intend to pattern match on this, it shouldn't be a case class. Case classes cause serious difficulties when trying to evolve the API in compatible ways.

File: tools/src/main/scala/scala/scalanative/io/Classpath.scala (hunk 1 above)
Comment (densh) [inline tools/src/main/scala/scala/scalanative/io/Classpath.scala]: Also, do we really need to know a precise type `T` here? It seems that `VirtualFile` should be enough.

File: tools/src/main/scala/scala/scalanative/io/Classpath.scala [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  override def close(): Unit = ()
+
+  override def contents: Stream[LazyVFile[LocalClasspath]] = Files.walk(path,Integer.MAX_VALUE, FileVisitOption.FOLLOW_LINKS)
+    .iterator()
+    .asScala.toStream
+    .map(filePath => LazyVFile[LocalClasspath](filePath, this))
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/io/Classpath.scala]: The very first classpath traversal will force this. We should just use a `Seq` and compute this eagerly. Scala `Stream` is very unpredictable performance-wise.

File: tools/src/main/scala/scala/scalanative/io/Classpath.scala (hunk 4 above)
Comment (densh) [inline tools/src/main/scala/scala/scalanative/io/Classpath.scala]: Given that we only have LazyVirtualFiles, why even bother VirtualFile/LazyVirtualFile distinction. We don't want to prematurely abstract things.

Comment (densh) [inline tools/src/main/scala/scala/scalanative/io/Classpath.scala]: Lastly, value of type parameter here is not clear.

File: tools/src/main/scala/scala/scalanative/linker/Linker.scala [hunk 8]
```diff
@@ -42,6 +42,8 @@ final class Linker(dotpath: Option[String], paths: Seq[String]) {
     }
 
   def link(entry: Global): (Seq[Global], Seq[Defn]) = {
+    println(s"Linking $entry")
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/linker/Linker.scala]: This and all the other `println`-s in Linker need to be removed. We need to think of proper logging implementation instead (https://github.com/scala-native/scala-native/issues/119) but that's out of scope for this pr.

File: tools/src/main/scala/scala/scalanative/io/Classpath.scala (hunk 1 above)
Comment (Milyardo) [inline tools/src/main/scala/scala/scalanative/io/Classpath.scala]: Sorry about that, noise left over from an approach where classpath was a typeclass.

**Comment by Milyardo** [comment]
Would you like the history to be squashed at this point?

File: tools/src/main/scala/scala/scalanative/linker/Linker.scala [hunk 9]
```diff
@@ -42,6 +42,7 @@ final class Linker(dotpath: Option[String], paths: Seq[String]) {
     }
 
   def link(entry: Global): (Seq[Global], Seq[Defn]) = {
+
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/linker/Linker.scala]: No need for this line.

File: tools/src/main/scala/scala/scalanative/linker/Classpath.scala [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+  * A memory cached file-like object contained in some member of a classpath.
+  */
+
+final class VirtualFile(val relativePath: Path) {
+  def uri: URI = relativePath.toUri
+  lazy val bytes: ByteBuffer = ByteBuffer.wrap(Files.readAllBytes(relativePath))
+}
+
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/linker/Classpath.scala]: No need for this line.

File: tools/src/main/scala/scala/scalanative/linker/Assembly.scala [hunk 11]
```diff
@@ -2,40 +2,20 @@ package scala.scalanative
...
-  final case class Dir private[Assembly](val base: File) extends Assembly {
-    //println(s"discovered dir assembly $base")
+import scala.scalanative.nir._
+import scala.scalanative.nir.serialization._
 
-    class NirMatcher extends BiPredicate[Path, BasicFileAttributes] {
-      def test(path: Path, b: BasicFileAttributes) =
-        path.toString.endsWith(".nir")
-    }
+final case class Assembly(base: File) {
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/linker/Assembly.scala]: And this one.

**Comment by densh** [comment]
Squash, remove extra unnecessary blank lines, reformat with scalafmt, sign Scala CLA and it's good to merge! 👍

**Comment by densh** [comment]
@Milyardo It looks like your rebase went wrong. First commit has diff that changes code all over the place.

**Pull request closed without merging**